pkg/client-nfd/compat/node-validator/node-validator.go

```diff
@@ -96,91 +97,100 @@ func (nv *nodeValidator) Execute(ctx context.Context) ([]*CompatibilityStatus, e
 	return compats, nil
 }
 
-func evaluateRuleStatus(rule *nfdv1alpha1.Rule, matchStatus *nodefeaturerule.MatchStatus) ProcessedRuleStatus {
-	var matchedFeatureTerms nfdv1alpha1.FeatureMatcher
+func (nv *nodeValidator) evaluateRuleStatus(rule *nfdv1alpha1.Rule, matchStatus *nodefeaturerule.MatchStatus) ProcessedRuleStatus {
 	out := ProcessedRuleStatus{Name: rule.Name, IsMatch: matchStatus.IsMatch}
 
-	evaluateFeatureMatcher := func(featureMatcher, matchedFeatureTerms nfdv1alpha1.FeatureMatcher) []MatchedExpression {
-		out := []MatchedExpression{}
-		for _, term := range featureMatcher {
-			if term.MatchExpressions != nil {
-				for name, exp := range *term.MatchExpressions {
-					isMatch := false
-
-					// Check if the expression matches
-					for _, processedTerm := range matchedFeatureTerms {
-						if term.Feature != processedTerm.Feature || processedTerm.MatchExpressions == nil {
-							continue
-						}
-						pexp, ok := (*processedTerm.MatchExpressions)[name]
-						if isMatch = ok && exp.Op == pexp.Op && slices.Equal(exp.Value, pexp.Value); isMatch {
-							break
-						}
-					}
-
-					out = append(out, MatchedExpression{
-						Feature:     term.Feature,
-						Name:        name,
-						Expression:  exp,
-						MatcherType: MatchExpressionType,
-						IsMatch:     isMatch,
-					})
-				}
-			}
+	matchedFeatureTerms := nfdv1alpha1.FeatureMatcher{}
+	if matchStatus.MatchFeatureStatus != nil {
+		matchedFeatureTerms = matchStatus.MatchFeatureStatus.MatchedFeaturesTerms
+	}
+	out.MatchedExpressions = nv.matchFeatureExpressions(rule.MatchFeatures, matchedFeatureTerms)
 
-			if term.MatchName != nil {
-				isMatch := false
-				for _, processedTerm := range matchStatus.MatchedFeaturesTerms {
-					if term.Feature != processedTerm.Feature || processedTerm.MatchName == nil {
-						continue
-					}
-					isMatch = term.MatchName.Op == processedTerm.MatchName.Op && slices.Equal(term.MatchName.Value, processedTerm.MatchName.Value)
-					if isMatch {
-						break
-					}
-				}
-				out = append(out, MatchedExpression{
-					Feature:     term.Feature,
-					Name:        "",
-					Expression:  term.MatchName,
-					MatcherType: MatchNameType,
-					IsMatch:     isMatch,
-				})
-			}
+	for i, matchAnyElem := range rule.MatchAny {
+		matchedFeatureTermsAny := nfdv1alpha1.FeatureMatcher{}
+		if matchStatus.MatchAny[i].MatchedFeaturesTerms != nil {
+			matchedFeatureTermsAny = matchStatus.MatchAny[i].MatchedFeaturesTerms
```

Comment:
nit: just caught my eye, small readability improvement (imo)

```suggestion
		if t := matchStatus.MatchAny[i].MatchedFeaturesTerms; t != nil {
			matchedFeatureTermsAny = t
```

Reply:
Fixed.